README.md
# RDictCc

RDictCc is a simple translation tool written in Ruby that creates a local
database out of the dictionary files from [dict.cc](http://www.dict.cc).

## Usage

Simply place the file `rdictcc.rb` somewhere in your `PATH`.  Executing
`rdictcc.rb --help` will show you all that's needed.

To make use of the Emacs mode, add this to your `~/.emacs`.

```
;; Adapt path as you need
(add-to-list 'load-path "~/path/to/rdictcc")
(setq rdictcc-program "~/path/to/rdictcc/rdictcc.rb")

(require 'rdictcc)
;; Adapt to your likings
(global-set-key (kbd "C-c t") 'rdictcc-translate-word-at-point)
(global-set-key (kbd "C-c T") 'rdictcc-translate-word)

```

Have fun!

## License

Copyright (C) 2012 Tassilo Horn <tassilo@member.fsf.org>

Distributed under the General Public License, Version 3.


ChangeLog
2007-09-05  Tassilo Horn  <tassilo@member.fsf.org>

	* rdictcc.el (rdictcc-pop-to-translation-buffer): Use
	inhibit-read-only instead of buffer-read-only.
	(rdictcc-translate-word): Don't use word-at-point but
	rdictcc-current-word and use default value except initial-input.

2007-08-26  Tassilo Horn  <tassilo@member.fsf.org>

	* rdictcc.el (rdictcc-translate-word-to-string): Set
	coding-system-for-{read,write} to terminal-coding-system.



rdictcc.el
;;; rdictcc.el --- use the rdictcc.rb client from within Emacs

;; Copyright (C) 2006, 2007 by Tassilo Horn

;; Author: Tassilo Horn <tassilo@member.fsf.org>

;; This program is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by the Free
;; Software Foundation; either version 3, or (at your option) any later
;; version.

;; This program is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
;; FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
;; more details.

;; You should have received a copy of the GNU General Public License along with
;; this program ; see the file COPYING.  If not, write to the Free Software
;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

;;; Commentary:

;; Some functions to let you use dictcc.rb from within emacs.

(defgroup rdictcc
  nil
  "A client for accessing the rdictccserver. It provides a fast
and convenient German-English (and vice versa) translation."
  :group 'hypermedia)

(defcustom rdictcc-program
  "/usr/local/bin/rdictcc.rb"
  "The path to rdictcc.rb client app."
  :group 'rdictcc
  :type 'string)

(defcustom rdictcc-show-translations-in-tooltips
  nil
  "If set to t, translations will be shown in tooltips. Tooltips
are only available in GNU Emacs' X11 interface."
  :group 'rdictcc
  :type 'boolean)

(defcustom rdictcc-show-translations-in-buffer
  t
  "If set to t, translations will be shown in a separate
*rdictcc* buffer."
  :group 'rdictcc
  :type 'boolean)

(defvar rdictcc-last-word nil
  "The last translated word (internal use only)")

(defvar rdictcc-last-translation nil
  "The last translation (internal use only)")

(defvar rdictcc-version "<2007-09-05 Wed 11:30>"
  "rdictcc.el's version")

(defun rdictcc-translate-word-to-string (word)
  "Translates the given word and returns the result as string."
  (if (equal word rdictcc-last-word)
      rdictcc-last-translation
    (let* ((coding-system-for-read  (terminal-coding-system))
           (coding-system-for-write (terminal-coding-system))
           (translation (shell-command-to-string
                         (concat rdictcc-program " " word))))
      (setq rdictcc-last-word word)
      (setq rdictcc-last-translation translation))))

(defun rdictcc-translate-word (word)
  "Asks the use for a word and translates it with
`rdictcc-program'. The result will be displayed in a window with
*rdictcc* buffer or/and in a tooltip. The variables
`rdictcc-show-translations-in-buffer' and
`rdictcc-show-translations-in-tooltips' influence this
behavior. The *rdictcc* buffer has his own major mode with useful
key bindings. Type `?' in it to get a description."
  (interactive (list (let ((inhibit-read-only t))
                       (substring-no-properties
                        (read-string (concat "Word to translate (defaults to \""
                                             (rdictcc-current-word)
                                             "\"): ")
                                     nil nil (rdictcc-current-word))))))
  (let ((translation (rdictcc-translate-word-to-string word)))
    (when rdictcc-show-translations-in-buffer
      (rdictcc-pop-to-translation-buffer translation))
    (when (and rdictcc-show-translations-in-tooltips window-system)
      (tooltip-show translation))))

(defvar rdictcc-old-window-configuration nil
  "The window configuration which has to be restored when the
*rdictcc* buffer is closed. (internal use only)")

(defun rdictcc-pop-to-translation-buffer (translation)
  (setq rdictcc-old-window-configuration (current-window-configuration))
  (pop-to-buffer "*rdictcc*" nil t)
  (rdictcc-buffer-mode)
  (setq inhibit-read-only t)
  (erase-buffer)
  (insert translation)
  (setq inhibit-read-only nil)
  (goto-char (point-min))
  (rdictcc-next-translation))

(defun rdictcc-current-word ()
  (if (>= emacs-major-version 22)
      (current-word t t) ; emacs 22
    (current-word t)))   ; emacs 21

(defun rdictcc-translate-word-at-point ()
  "Opens a new window showing the translation of the word located
at point."
  (interactive)
  (save-excursion
    (let ((word (rdictcc-current-word)))
      (rdictcc-translate-word word))))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; RdictCc Buffer Mode (major mode)

(define-derived-mode rdictcc-buffer-mode nil "rdictcc buffer"
  "The buffer used in the *rdictcc* buffers. Has some convenient
key bindings to allow fast usage, and the word whose translations
is displayed is highlighted with `font-lock-keyword-face'."
  (setq buffer-read-only t) ; *rdictcc* is read-only by default
  ;;;;;;;;;;;;;;;
  ;; Keymap Stuff
  (suppress-keymap rdictcc-buffer-mode-map)
  (define-key rdictcc-buffer-mode-map (kbd "q") 'rdictcc-close-buffer)
  (define-key rdictcc-buffer-mode-map (kbd "o") 'other-window)
  (define-key rdictcc-buffer-mode-map (kbd "RET")
    'rdictcc-replace-word-with-translation-at-point)
  (define-key rdictcc-buffer-mode-map (kbd "?") 'describe-mode)
  (define-key rdictcc-buffer-mode-map (kbd "n") 'rdictcc-next-translation)
  (define-key rdictcc-buffer-mode-map (kbd "p") 'rdictcc-previous-translation)
  (define-key rdictcc-buffer-mode-map (kbd "W d")
    'rdictcc-webtranslate-last-word-with-dictcc)
  (define-key rdictcc-buffer-mode-map (kbd "W l")
    'rdictcc-webtranslate-last-word-with-leo)
  ;;;;;;;;;;;;;;;;;;
  ;; Font Lock Stuff
  (make-local-variable 'font-lock-defaults)
  (setq font-lock-defaults
        ;; The translated word is highlighted in *rdictcc* buffer.
        `((,(concat "\\b" rdictcc-last-word "\\b")) t t)))

(defun rdictcc-replace-word-with-translation-at-point ()
  "Replaces the translated word with the translation at point in
*rdictcc* buffer."
  (interactive)
  (let ((chosen-translation (rdictcc-current-word)))
    (rdictcc-close-buffer)
    (search-forward-regexp "\\>")
    (backward-kill-word 1)
    (insert chosen-translation)))

(defun rdictcc-close-buffer ()
  "Closes the *rdictcc* buffer and restores the window
configuration which existed before the translation."
  (interactive)
  (set-window-configuration rdictcc-old-window-configuration))

(defun rdictcc-next-translation ()
  "Go to the next translation in *rdictcc* buffer."
  (interactive)
  (search-forward-regexp (concat "^.*" rdictcc-last-word ".*:") nil t))

(defun rdictcc-previous-translation ()
  "Go to the previous translation in *rdictcc* buffer."
  (interactive)
  (search-backward-regexp (concat "^.*" rdictcc-last-word ".*:") nil t))

(defvar rdictcc-webtranslate-symbols '("dictcc" "leo")
  "A list of symbol names (as strings) which name the translation
sites rdictcc can use as webtranslation sites. These strings are
used to complete the site symbol when interactively calling
`rdictcc-webtranslate'.")

(defun rdictcc-webtranslate-last-word-with-dictcc ()
  "Openes the browser specified by `browse-url-browser-function'
with http://www.dict.cc's translation of the last translated
word."
  (interactive)
  (rdictcc-webtranslate rdictcc-last-word 'dictcc))

(defun rdictcc-webtranslate-last-word-with-leo ()
  "Openes the browser specified by `browse-url-browser-function'
with http://dict.leo.org's translation of the last translated
word."
  (interactive)
  (rdictcc-webtranslate rdictcc-last-word 'leo))

(defun rdictcc-webtranslate (word &optional site)
  "Opens the browser defined by `browse-url-browser-function' and
translates the given word at the given site, which has to be one
symbol of

    * dictcc or (none): Translation at http://www.dict.cc

    * leo: Translation at http://dict.leo.org

If no symbol is given http://dict.cc will be queried."
  (interactive
   (list (read-string "Word to translate: ")
         (completing-read "Site symbol: " rdictcc-webtranslate-symbols)))
  (cond
   ((eq site 'leo)
    (browse-url (concat "http://dict.leo.org/?search=" word)))
   (t
    (browse-url (concat "http://www.dict.cc/?s=" word)))))

;;; end of RdictCc Buffer Mode
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Tooltip mode stuff (minor mode)

(defvar rdictcc-tooltip-mode nil
  "Indicates wheather the rdictcc tooltip mode is active. Setting
this variable doesn't have any effect. Use function
`rdictcc-tooltip-mode' instead.")
(nconc minor-mode-alist '((rdictcc-tooltip-mode " RDictCc")))

(defcustom rdictcc-tooltip-delay
  2
  "How long should the mouse be over a word until the translation
will be displayed in a tooltip. Don't set it to a too low value."
  :group 'rdictcc
  :type 'number)

(defun rdictcc-translate-word-open-tooltip (event)
  "Display translations of the word under the mouse pointer in a
tooltip."
  (interactive "e")
  (let ((word (save-window-excursion
                (save-excursion
                  (mouse-set-point event)
                  (rdictcc-current-word)))))
    (when word
      (if (string= word rdictcc-last-word)
          (tooltip-show rdictcc-last-translation)
        (tooltip-show (rdictcc-translate-word-to-string word))))))

(defun rdictcc-tooltip-mode (&optional arg)
  "Display tooltips with the translations of the word under the
mouse pointer."
  (interactive "P")
  (require 'tooltip)
  (require 'gud) ;; The tooltips with events are currently part of GUD
  (let ((val (if arg
                 (> (prefix-numeric-value arg) 0)
               (not rdictcc-tooltip-mode))))
    (if val
        ;; Switch tooltip mode on
        (progn
          (make-local-variable 'rdictcc-tooltip-mode)
          (setq rdictcc-tooltip-mode val)
          (make-local-variable 'tooltip-delay)
          (setq tooltip-delay rdictcc-tooltip-delay)
          (gud-tooltip-mode 1)
          (add-hook 'tooltip-hook 'rdictcc-translate-word-open-tooltip t t)
          (make-local-variable 'track-mouse)
          (setq track-mouse val))
      ;; Switch tooltip mode off
      (kill-local-variable 'rdictcc-tooltip-mode)
      (kill-local-variable 'tooltip-delay)
      (kill-local-variable 'track-mouse)
      (remove-hook 'tooltip-hook 'rdictcc-translate-word-open-tooltip t))))

;;; end of tooltip stuff
;;;;;;;;;;;;;;;;;;;;;;;;

(provide 'rdictcc)

;;; rdictcc ends here


rdictcc.rb
#!/usr/bin/ruby -w

###############################################################################
# Copyright (C) 2006, 2007 by Tassilo Horn
#
# Author: Tassilo Horn <tassilo@member.fsf.org>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program ; see the file COPYING.  If not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
###############################################################################

require 'optparse' # Command line option parsing
require 'dbm'      # DBM database interface

##
# Constants used by all classes
DICT_DIR = File.expand_path '~/.rdictcc'
DICT_FILE_DE = DICT_DIR + '/' + 'dict_de'
DICT_FILE_EN = DICT_DIR + '/' + 'dict_en'


##
# A RDictCcEntry contains a Hash
#
#     phrase => [translation1, ..., translationN]
#
# and is heavily used while importing the Dict.cc CSV-file. The String returned
# by to_s() encodes a RDictCcEntry as string, which becomes the value of some
# keyword in the DBM database. After importing the Dict.cc CSV-file to DBM
# database files the only method used is the static format_str(str), which
# takes a string encoded RDictCcEntry (DBM-value) and formats them
# userfriendly.
class RDictCcEntry
  def initialize
    @value_hash = {}
  end

  ##
  # Append 'phrase' and 'translation'. If this 'phrase' already exists as key,
  # then simply append 'translation' to the value array.
  def append(phrase, translation)
    phrase.strip!
    translation.strip!
    if @value_hash[phrase].nil?
      @value_hash[phrase] = [translation]
    else
      @value_hash[phrase] << translation
    end
  end

  ##
  # Encodes this RDictCcEntry as string which is used to store entries as
  # values of the DBM database.
  def to_s
    s = ""
    # The results should be listed from shortest (exact) to longest match. So
    # we store it that way.
    ary = @value_hash.sort { |a, b| a[0].size <=> b[0].size }
    ary.each do |elem|
      s << "#{elem[0]}=<>"
      elem[1].each do |val|
        s << "#{val}:<>:"
      end
      s.gsub!(/:<>:$/, '#<>#')
    end
    s << "\n"
    s
  end

  ##
  # Given a string-encoded (with to_s()) RDictCcEntry formats it in a readable,
  # userfriendly way.
  def RDictCcEntry.format_str( str )
    parts = str.strip!.split(/#<>#/)
    s = ""
    parts.each do |part|
      subparts = part.split(/=<>/)
      s << subparts[0] + ":\n" # That's the description
      subparts[1].split(/:<>:/).each { |trans| s << "    - " + trans + "\n" }
    end
    s
  end
end

##
# When the '-i | --import' is given this class builds the DBM files out of the
# textual dict file you can download at http://www.dict.cc
class RDictCcDatabaseBuilder
  ##
  # Creates a new DatabaseBuilder, initializes it with the textual dict file
  # and creates the ~/.rdictcc directory.
  def initialize( import_file )
    @import_file = import_file
    if !File.exists? DICT_DIR
      Dir.mkdir DICT_DIR
    end
  end

  ##
  # Imports the dict.cc file given in the constructor and builds/writes the
  # database files.
  def import
    # German => English
    read_dict_file(:de)
    write_database(:de)
    # English => German
    read_dict_file(:en)
    write_database(:en)
  end

  ##
  # Below go the private things...
  private

  ##
  # Writes the contents of '@dict' to DBM file DICT_FILE_DE if sym == :de or
  # to DICT_FILE_EN otherwise.
  def write_database( sym )
    if sym == :de
      db_file = DICT_FILE_DE
    else
      db_file = DICT_FILE_EN
    end

    if File.exists?(db_file)
      puts "** Going to delete old database #{db_file}"
      File.delete(db_file)
      puts "** Deleted old database #{db_file}"
    end
    # Write to db
    puts "** writing DBM database file..."
    DBM.open(db_file, 0644, DBM::NEWDB) do |dbm|
      i = 0
      @dict.each_pair do |keyword, value|
        dbm[keyword] = value.to_s
        i += 1
        puts "Stored #{i} / #{@dict.size} values" if i % 1000 == 0
      end
    end

    # Now @dict is useless, so get rid of it
    @dict = nil
    GC.start
    puts "** Database building done!"
  end

  ##
  # Builds the '@dict' Hash from 'dict_file'. If symbol 'sym' is :de, the
  # DE-EN-Dictionary will be build, else the EN-DE is build.
  def read_dict_file( sym )
    @dict = {}
    # No queries allowed until reading finishes
    puts "** Reading dict file (#{sym})"
    IO.foreach(@import_file) do |line|
      line.strip!
      # skip empty lines and comments
      next if line =~ /^#/ or line =~ /^\s$/ or line.empty?

      # add the line to the dict
      add_line(line, sym)
    end
    puts "** built dict with #{@dict.size} entries"
  end

  ##
  # Add the line 'str' to '@dict'.
  def add_line( str, sym )
    # split the line
    if sym == :de
      phrase, translation = str.split('::')
    else
      translation, phrase = str.split('::')
    end
    word = extract_word(phrase)

    ## debug
    # puts word
    ## debug

    # Add another entry
    if !word.nil?
      @dict[word] ||= RDictCcEntry.new
      @dict[word].append(phrase, translation)
    end
  end

  ##
  # Cause the CSV-file contains phrases we cannot be sure what is the most
  # important word. This method strikes out everything between parenthesis, and
  # if multiple words stay over, simply takes the longes one.
  def extract_word( phrase )
    w = phrase.gsub(/(\([^(]*\)|\{[^{]*\}|\[[^\[]*\])/, '').strip.downcase
    return nil if w.empty? # No empty strings
    # Now return the longest word, hoping that it's the most important, too
    ary = w.gsub(/[^üäöß\w\s-]/, '').split # öäüß are no word chars currently!
    ary.sort!{ |x,y| y.length <=> x.length }
    ary[0] # The longest element is the first
  end
end

class RDictCcQueryEvaluator
  def initialize
    if !File.exists? DICT_DIR
      puts "There's no ~/.rdictcc directory! You have to import an dict.cc\n" +
        "database file first. See\n" +
        "  $ rdictcc.rb --help\n" +
        "for more information."
      exit
    end
  end

  ##
  # Opens each database and yields the given block, handing over the data base
  # handle.
  def read_db
    for file in [DICT_FILE_DE, DICT_FILE_EN] do
      if file == DICT_FILE_DE
        puts "DE-EN Translations:"
        puts "==================="
      else
        puts "\nEN-DE Translations:"
        puts "==================="
      end

      DBM.open(file, nil, DBM::READER) do |dbm|
        yield dbm
      end
    end
  end

  ##
  # Delegates queries according to query type.
  def query( query )
    query.downcase!
    case query
    when /^:r:/ then query_regexp query.gsub(/^:r:/, '')
    when /^:f:/ then query_fulltext_regexp query.gsub(/^:f:/, '')
    else query_simple query
    end
  end

  ##
  # Simple hash lookup. Complexity: O(1)
  def query_simple( query )
    read_db do |dbm|
      puts RDictCcEntry.format_str(dbm[query]) if !dbm[query].nil?
    end
  end

  ##
  # Regexp lookup. Complexity: O(n)
  def query_regexp( query )
    read_db do |dbm|
      dbm.each_key do |key|
        puts RDictCcEntry.format_str(dbm[key]) if key =~ /#{query}/
      end
    end
  end

  ##
  # Fulltext regexp lookup. Complexity: O(n)
  def query_fulltext_regexp( query )
    read_db do |dbm|
      dbm.each_value do |raw_val|
        val = RDictCcEntry.format_str(raw_val)
        match_line_found = false
        val.each_line do |line|
          if line =~ /^\s+/
            if match_line_found
              puts line
            else
              # Skip lines starting with blanks, because these are already
              # translations and they don't belong to the matching line.
              next
            end
          else
            match_line_found = false
          end
          if line.downcase =~ /#{query}/
            puts line
            match_line_found = true
          end
        end
      end
    end
  end

  def show_db_sizes
    read_db do |dbm|
      i = 0
      # That's probably faster and less memory consuming than
      # dbm.entries.size...
      dbm.each { |e| i += 1 }
      puts "Database has #{i} entries"
    end
  end
end

def interactive_mode
  puts "Welcome to rdictcc's interactive mode. This mode will read from stdin\n" +
    "and print the translations until it reads ^Q (literally, not Ctrl-Q)."
  evaluator = RDictCcQueryEvaluator.new
  print "=> "
  while word = gets.chomp
    puts
    if word == "^Q" then break end
    evaluator.query word
    puts "--------------------------------------------------------------------------------"
    print "=> "
  end
  puts "Bye."
  exit 0
end

##
# Here we go...
$query_str = ""
options = OptionParser.new do |opts|
  opts.banner = "Usage: rdictcc.rb [database_import_options]\n" +
    "       rdictcc.rb [misc_options]\n" +
    "       rdictcc.rb [query_option] QUERY\n" +
    "       rdictcc.rb\n"

  opts.separator ""
  opts.separator "If no option nor QUERY is given, you'll enter rdictcc's interactive mode."

  opts.separator ""
  opts.separator "Database building options:"
  opts.on("-i", "--import DICTCC_FILE",
          "Import the dict file from dict.cc") do |file|
    db_builder = RDictCcDatabaseBuilder.new(file)
    db_builder.import
    exit 0
  end

  opts.separator ""
  opts.separator "Misc options:"
  opts.on("-v", "--version", "Show rdictcc.rb's version") do
    # TODO: Set version after changes!
    puts "<2007-08-08 Wed 18:50>"
    exit 0
  end

  opts.on("-S", "--size", "Show the number of entries in the databases") do
    RDictCcQueryEvaluator.new.show_db_sizes
    exit 0
  end

  opts.on("-h", "--help", "Show this message") do
    puts opts
    exit 0
  end

  opts.separator ""
  opts.separator "Query option:"
  opts.on("-s", "--simple", "Translate the word given as QUERY (default)") do
    # No need to do anything...
  end

  opts.on("-r", "--regexp", "Translate all words matching the regexp QUERY") do
    $query_str = ":r:"
  end

  opts.on("-f", "--fulltext", "Translate all sentences matching the regexp QUERY") do
    $query_str = ":f:"
  end
end.parse!

# catch queries without QUERY
if ARGV.join(" ").empty?
  interactive_mode
end

evaluator = RDictCcQueryEvaluator.new
evaluator.query $query_str.concat(ARGV.join(" "))
